// @vercel/react-router/edge/entry.server.js
import { isbot } from 'isbot';
import { createElement } from 'react';
import { renderToReadableStream, } from 'react-dom/server';
import { ServerRouter } from 'react-router';
const vercelDeploymentId = process.env.VERCEL_DEPLOYMENT_ID;
const vercelSkewProtectionEnabled = process.env.VERCEL_SKEW_PROTECTION_ENABLED === '1';
export async function handleRequest(request, responseStatusCode, responseHeaders, routerContext, _loadContext, options) {
    const body = await renderToReadableStream(createElement(ServerRouter, {
        context: routerContext,
        url: request.url,
        nonce: options?.nonce,
    }), {
        ...options,
        signal: request.signal,
        onError(error) {
            console.error(error);
            responseStatusCode = 500;
        },
    });
    if (isbot(request.headers.get('user-agent'))) {
        await body.allReady;
    }
    responseHeaders.set('Content-Type', 'text/html');
    if (vercelSkewProtectionEnabled && vercelDeploymentId) {
        responseHeaders.append('Set-Cookie', `__vdpl=${vercelDeploymentId}; HttpOnly`);
    }
    return new Response(body, {
        headers: responseHeaders,
        status: responseStatusCode,
    });
}
